src/CommunityPatch/Patches/Perks/Intelligence/Engineering/ResolutePatch.cs

```diff
@@ -0,0 +1,165 @@
+      return additiveMoralePenaltyReductionRate;
+    }
+
+    private static float CalculateMultiplicativeBonus(float partyMemberBonus, int partyMemberCount) {
+      var finalRate = 0f;
+      var lastRate = 1f;
+      
+      for (var i = 0; i < partyMemberCount; i++) {
+        var nextRate = partyMemberBonus * lastRate;
+        finalRate += nextRate;
+        lastRate = nextRate;
+      }
+
+      return finalRate;
+    }
+
+    private static int CalculateTotalMoraleLossReduction(DefaultPartyMoraleModel model, MobileParty party, float reductionRate)
+    {
+      var recentEventsMoraleLossReduction = Math.Min(party.RecentEventsMorale, 0);
+      var starvationMoraleLossReduction = CalculateStarvationMoralePenalty(model, party);
+      var unpaidWageMoraleLossReduction = CalculateUnpaidWageMoralePenalty(model, party);
+      var partySizeMoraleLossReduction = CalculatePartySizeMoralePenalty(model, party);
+      var foodVarietyMoraleLossReduction = CalculateFoodVarietyPenalty(model, party);
+
+      var totalMoraleLossReduction = recentEventsMoraleLossReduction +
+        starvationMoraleLossReduction +
+        unpaidWageMoraleLossReduction +
+        partySizeMoraleLossReduction +
+        foodVarietyMoraleLossReduction;
+      
+      return (int) (totalMoraleLossReduction * reductionRate);
+    }
+
+    private static float CalculateUnpaidWageMoralePenalty(DefaultPartyMoraleModel model, MobileParty mobileParty)
+      => GetMoralePenaltyFromMethod(model, "GetNoWageMoralePenalty", mobileParty) * mobileParty.HasUnpaidWages;
+
+    private static float CalculateStarvationMoralePenalty(DefaultPartyMoraleModel model, MobileParty mobileParty)
+      => mobileParty.Party.IsStarving ? GetMoralePenaltyFromMethod(model, "GetStarvationMoralePenalty", mobileParty) : 0f;
+
+    private static float CalculatePartySizeMoralePenalty(DefaultPartyMoraleModel model, MobileParty party)
+      => GetMoralePenaltyFromMethodWithExplainedNumber(model, "GetPartySizeMoraleEffect", party);
+
+    private static float CalculateFoodVarietyPenalty(DefaultPartyMoraleModel model, MobileParty party)
+      => GetMoralePenaltyFromMethodWithExplainedNumber(model, "CalculateFoodVarietyMoraleBonus", party);
+
+    private static float GetMoralePenaltyFromMethod(DefaultPartyMoraleModel model, string methodName, MobileParty party) {
+      var method = DefaultPartyMoraleModelType.GetMethod(methodName, BindingFlags.Instance | BindingFlags.DeclaredOnly | BindingFlags.NonPublic);
+      if (method == null) return 0f;
+      
+      var moraleBonus = (int) method.Invoke(model, new object[] {party});
+      return GetMoralePenalty(moraleBonus);
+    }
+    
+    private static float GetMoralePenaltyFromMethodWithExplainedNumber(DefaultPartyMoraleModel model, string methodName, MobileParty party) {
+      var method = DefaultPartyMoraleModelType.GetMethod(methodName, BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.DeclaredOnly);
```

Comment:
Instead of passing the method name as string every call, this function could receive a `MethodInfo` and just invoke it. So you'd declere the MethodInfo just once:
`private static readonly GetNoWageMoralePenaltyMethodInfo = DefaultPartyMoraleModelType.GetMethod(methodName, BindingFlags.Instance | BindingFlags.DeclaredOnly | BindingFlags.NonPublic);`
and then the method signature would be `GetMoralePenaltyFromMethodWithExplainedNumber(DefaultPartyMoraleModel model, MethodInfo method, MobileParty party)` and you could just `method.Invoke(model, new object[] {party})`.
All this is doing is preventing a `GetMethod` call every time this function is called, so I'm not sure if it's worth it.
What do you think, @Tyler-IN?

Reply:
Yeah, need to extract as much reflection out to members, either static or instance.

Reply:
Will do this on dev. Need to get an RC going.